File: sdk/node/test/metrics.test.ts
import assert from "node:assert";
import { Agent, InMemoryBlobStore } from "../src/runtime";
import { loadRoot } from "../src/protos";
import { encodeDeterministic } from "../src/codec";
import type { MetricsHook } from "../src/metrics";
import { noopMetrics } from "../src/metrics";
import type { Msg, Subscription, SubscriptionOptions } from "nats";

class MockNatsConnection {
  published: Array<{ subject: string; data: Uint8Array }> = [];
  subscriptions: Map<string, NonNullable<SubscriptionOptions["callback"]>> = new Map();

  async publish(subject: string, data: Uint8Array): Promise<void> {
    this.published.push({ subject, data });
  }

  subscribe(subject: string, opts: SubscriptionOptions = {}): Subscription {
    if (opts.callback) {
      this.subscriptions.set(subject, opts.callback);
    }
    return { drain: async () => {} } as unknown as Subscription;
  }

  async drain(): Promise<void> {
    return;
  }
}

class RecordingMetrics implements MetricsHook {
  received: Array<{ jobId: string; topic: string }> = [];
  completed: Array<{ jobId: string; status: string }> = [];
  failed: Array<{ jobId: string; errorMsg: string }> = [];

  onJobReceived(jobId: string, topic: string): void {
    this.received.push({ jobId, topic });
  }
  onJobCompleted(jobId: string, _durationMs: number, status: string): void {
    this.completed.push({ jobId, status });
  }
  onJobFailed(jobId: string, errorMsg: string): void {
    this.failed.push({ jobId, errorMsg });
  }
  onHeartbeatSent(_workerId: string): void {}
}

async function waitFor(fn: () => boolean, timeoutMs = 1000): Promise<void> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (fn()) return;
    await new Promise((resolve) => setTimeout(resolve, 5));
  }
  throw new Error("timeout waiting for condition");
}

describe("MetricsHook", () => {
  it("receives OnJobReceived + OnJobCompleted on success", async () => {
    const store = new InMemoryBlobStore();
    const mock = new MockNatsConnection();
    const metrics = new RecordingMetrics();
    const jobId = "metrics-1";
    const ctxKey = `ctx:${jobId}`;
    await store.set(ctxKey, Buffer.from(JSON.stringify({ prompt: "test" }), "utf-8"));

    const agent = new Agent({
      store,
      connectFn: async () => mock as any,
      senderId: "worker-m1",
      metrics,
    });

    agent.job("job.metrics", async (_ctx, data: any) => {
      return { summary: data.prompt };
    });

    await agent.start();

    const root = await loadRoot();
    const BusPacket = root.lookupType("cordum.agent.v1.BusPacket");
    const packetObj = BusPacket.fromObject({
      traceId: "t1",
      senderId: "c1",
      protocolVersion: 1,
      createdAt: { seconds: Math.floor(Date.now() / 1000), nanos: 0 },
      jobRequest: { jobId, topic: "job.metrics", contextPtr: `redis://${ctxKey}` },
    });
    const data = encodeDeterministic(BusPacket, packetObj);

    const cb = mock.subscriptions.get("job.metrics")!;
    cb(null, { data } as Msg);

    await waitFor(() => metrics.completed.length > 0);

    assert.strictEqual(metrics.received.length, 1);
    assert.strictEqual(metrics.received[0].jobId, jobId);
    assert.strictEqual(metrics.received[0].topic, "job.metrics");
    assert.strictEqual(metrics.completed.length, 1);
    assert.strictEqual(metrics.completed[0].jobId, jobId);
    assert.strictEqual(metrics.completed[0].status, "SUCCEEDED");
    assert.strictEqual(metrics.failed.length, 0);
  });

  it("receives OnJobFailed on failure", async () => {
    const store = new InMemoryBlobStore();
    const mock = new MockNatsConnection();
    const metrics = new RecordingMetrics();
    const jobId = "metrics-2";

    const agent = new Agent({
      store,
      connectFn: async () => mock as any,
      senderId: "worker-m2",
      metrics,
    });

    agent.job("job.metrics.fail", async (_ctx, _data: any) => {
      return {};
    });

    await agent.start();

    const root = await loadRoot();
    const BusPacket = root.lookupType("cordum.agent.v1.BusPacket");
    // Use a missing context key to trigger failure
    const packetObj = BusPacket.fromObject({
      traceId: "t2",
      senderId: "c2",
      protocolVersion: 1,
      createdAt: { seconds: Math.floor(Date.now() / 1000), nanos: 0 },
      jobRequest: { jobId, topic: "job.metrics.fail", contextPtr: "redis://missing-key" },
    });
    const data = encodeDeterministic(BusPacket, packetObj);

    const cb = mock.subscriptions.get("job.metrics.fail")!;
    cb(null, { data } as Msg);

    await waitFor(() => metrics.failed.length > 0);

    assert.strictEqual(metrics.received.length, 1);
    assert.strictEqual(metrics.failed.length, 1);
    assert.strictEqual(metrics.failed[0].jobId, jobId);
    assert.strictEqual(metrics.completed.length, 0);
  });

  it("noopMetrics has zero overhead", () => {
    noopMetrics.onJobReceived("j1", "t1");
    noopMetrics.onJobCompleted("j1", 100, "SUCCEEDED");
    noopMetrics.onJobFailed("j1", "err");
    noopMetrics.onHeartbeatSent("w1");
  });
});

File: sdk/go/runtime/metrics_test.go
package runtime

import (
	"context"
	"encoding/json"
	"sync"
	"testing"

	agentv1 "github.com/cordum-io/cap/v2/cordum/agent/v1"
	capsdk "github.com/cordum-io/cap/v2/sdk/go"
	"google.golang.org/protobuf/types/known/timestamppb"
)

type recordingMetrics struct {
	mu        sync.Mutex
	received  []string
	completed []string
	failed    []string
}

func (r *recordingMetrics) OnJobReceived(jobID, topic string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.received = append(r.received, jobID+":"+topic)
}

func (r *recordingMetrics) OnJobCompleted(jobID string, durationMs int64, status string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.completed = append(r.completed, jobID+":"+status)
}

func (r *recordingMetrics) OnJobFailed(jobID string, errorMsg string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.failed = append(r.failed, jobID+":"+errorMsg)
}

func (r *recordingMetrics) OnHeartbeatSent(workerID string) {}

func TestMetricsOnSuccess(t *testing.T) {
	store := NewInMemoryBlobStore()
	mock := newMockNATS()
	metrics := &recordingMetrics{}
	jobID := "metrics-job-1"
	ctxKey := "ctx:" + jobID
	payload, _ := json.Marshal(map[string]string{"prompt": "hello"})
	_ = store.Set(context.Background(), ctxKey, payload)

	agent := &Agent{
		HandshakeMode: HandshakeModeOff,
		NATS:          mock,
		Store:         store,
		SenderID:      "worker-m1",
		Metrics:       metrics,
	}

	Register(agent, "job.metrics", func(_ Context, input struct{ Prompt string }) (map[string]string, error) {
		return map[string]string{"summary": input.Prompt}, nil
	})

	if err := agent.Start(); err != nil {
		t.Fatalf("start: %v", err)
	}

	req := &agentv1.JobRequest{
		JobId:      jobID,
		Topic:      "job.metrics",
		ContextPtr: "redis://" + ctxKey,
	}
	packet := &agentv1.BusPacket{
		TraceId:         "trace-m1",
		SenderId:        "client-m1",
		CreatedAt:       timestamppb.Now(),
		ProtocolVersion: capsdk.DefaultProtocolVersion,
		Payload:         &agentv1.BusPacket_JobRequest{JobRequest: req},
	}
	data, _ := capsdk.MarshalDeterministic(packet)
	mock.send("job.metrics", data)

	metrics.mu.Lock()
	defer metrics.mu.Unlock()
	if len(metrics.received) != 1 || metrics.received[0] != jobID+":job.metrics" {
		t.Fatalf("expected OnJobReceived(%s, job.metrics), got %v", jobID, metrics.received)
	}
	if len(metrics.completed) != 1 || metrics.completed[0] != jobID+":SUCCEEDED" {
		t.Fatalf("expected OnJobCompleted(%s, SUCCEEDED), got %v", jobID, metrics.completed)
	}
	if len(metrics.failed) != 0 {
		t.Fatalf("expected no failures, got %v", metrics.failed)
	}
}

func TestMetricsOnFailure(t *testing.T) {
	store := NewInMemoryBlobStore()
	mock := newMockNATS()
	metrics := &recordingMetrics{}
	jobID := "metrics-job-2"
	// Empty context_ptr triggers failure
	agent := &Agent{
		HandshakeMode: HandshakeModeOff,
		NATS:          mock,
		Store:         store,
		SenderID:      "worker-m2",
		Metrics:       metrics,
	}

	Register(agent, "job.metrics.fail", func(_ Context, input struct{ Prompt string }) (map[string]string, error) {
		return map[string]string{"summary": input.Prompt}, nil
	})

	if err := agent.Start(); err != nil {
		t.Fatalf("start: %v", err)
	}

	req := &agentv1.JobRequest{
		JobId:      jobID,
		Topic:      "job.metrics.fail",
		ContextPtr: "redis://missing-key",
	}
	packet := &agentv1.BusPacket{
		TraceId:         "trace-m2",
		SenderId:        "client-m2",
		CreatedAt:       timestamppb.Now(),
		ProtocolVersion: capsdk.DefaultProtocolVersion,
		Payload:         &agentv1.BusPacket_JobRequest{JobRequest: req},
	}
	data, _ := capsdk.MarshalDeterministic(packet)
	mock.send("job.metrics.fail", data)

	metrics.mu.Lock()
	defer metrics.mu.Unlock()
	if len(metrics.received) != 1 {
		t.Fatalf("expected 1 OnJobReceived, got %d", len(metrics.received))
	}
	if len(metrics.failed) != 1 {
		t.Fatalf("expected 1 OnJobFailed, got %d", len(metrics.failed))
	}
	if len(metrics.completed) != 0 {
		t.Fatalf("expected 0 OnJobCompleted, got %d", len(metrics.completed))
	}
}
